### PR TITLE
Fix Lao Connection Bug

### PR DESCRIPTION
**Pull request by Warhedgehog**
This solves #948 

This is the implementation of a workaround where the change of websocket connection and the subscription/catch up process are done in separate activities. Therefore this adds a separate connection activity rather than a `HomeActivity` fragment.

The code smell is a false positive since the tests that makes no assertion, tests the opening of the `HomeActivity` upon pressing cancel.

### CODE REVIEW

**Review by Tesa27 — APPROVED** [review]

File: fe2-android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -8,17 +8,20 @@
...
     android:label="@string/app_name"
-    android:name="com.github.dedis.popstellar.PoPApplication"
     android:networkSecurityConfig="@xml/network_security_config"
     android:roundIcon="@mipmap/ic_launcher_round"
     android:supportsRtl="true"
     android:theme="@style/AppTheme">
     <activity
-      android:name="com.github.dedis.popstellar.ui.home.HomeActivity"
+      android:name=".ui.home.connecting.ConnectingActivity"
+      android:exported="false" />
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/AndroidManifest.xml:22]: I don't think the exported field is needed here.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java [hunk 2]
```diff
@@ -106,10 +106,12 @@ private void processIncomingMessages() {
   public Completable catchup(Channel channel) {
     Log.d(TAG, "sending a catchup to the channel " + channel);
     Catchup catchup = new Catchup(channel, requestCounter.incrementAndGet());
+
     return request(catchup)
         .map(ResultMessages.class::cast)
         .map(ResultMessages::getMessages)
         .doOnSuccess(messages -> Log.d(TAG, "Catchup on " + channel + " retrieved : " + messages))
+        .doOnError(msg -> Log.d(TAG, "Error in catchup is " + msg))
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java:114]: Isn't `doOnError()` supposed to provide a `Throwable` ?
And by design, we chose to keep the error handling to the caller. But this is a questionable choice.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java [hunk 3]
```diff
@@ -134,12 +136,9 @@ public Completable subscribe(Channel channel) {
...
         // Catchup already sent messages after the subscription to the channel is complete
-        // || TODO This should be used instead of the two next uncommented lines as it allows the
-        // || returned Completable to complete only when both subscribe and catchup are complete.
-        // || But right now, the LAO creation need this specific behavior.
-        // .flatMapCompletable(answer -> catchup(channel))
-        .doAfterSuccess(answer -> catchup(channel).subscribe())
-        .ignoreElement();
+        // This allows for the completion of the returned completable only when both subscribe
+        // and catchup are completed
+        .flatMapCompletable(answer -> catchup(channel));
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java:141]: The reason we used `doAfterSuccess()` is to catchup after the subscribe is finished. Because in the current app implementation, the Lao needs to be created before message are retrieved through the catchup. (which is a very questionable design choice 😅)

This change does fix a problem, but it creates another one that can be very tricky to debug as it introduce concurrency.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/MessageSender.java [hunk 4]
```diff
@@ -60,4 +60,6 @@ public interface MessageSender extends Disposable {
    * @return a {@link Completable} the will complete once the unsubscribe is finished
    */
   Completable unsubscribe(Channel channel);
+
+  Connection getConnection();
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/MessageSender.java:64]: Why do you need this ? We decided to abstract the connection away when implementing this part of the app. Is this really necessary ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java [hunk 5]
```diff
@@ -1181,8 +1181,8 @@ public void subscribeToLao(String laoId) {
                 lao -> {
                   Log.d(TAG, "got an update for lao: " + lao.getName());
                   mCurrentLao.postValue(lao);
+                  assert keyManager != null;
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java:1184]: Isn't `keyManager` injected through Hilt ?
When is this assert needed ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 6]
```diff
@@ -316,8 +317,10 @@ private void setupLaunchFragment() {
     setCurrentFragment(R.id.fragment_launch, LaunchFragment::newInstance);
   }
 
-  private void setupConnectingFragment() {
-    setCurrentFragment(R.id.fragment_connecting, ConnectingFragment::newInstance);
+  private void setupConnectingActivity(String laoId) {
+    Intent intent = new Intent(this, ConnectingActivity.class);
+    intent.putExtra("LAO_ID", laoId);
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:322]: I generally try to avoid "magic values" as much as possible.
Especially when this is used as a key.

Could you create a constant for this ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+  private ConnectingViewModel mViewModel;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.connecting_activity);
+    setConnectingFragment();
+
+    String channelId = (String) (getIntent().getExtras()).get("LAO_ID");
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java:34]: Unnecessary parenthesis

```suggestion
    String channelId = (String) getIntent().getExtras().get("LAO_ID");
```

Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java:34]: Unnecessary parenthesis

```suggestion
    String channelId = (String) getIntent().getExtras().get("LAO_ID");
```

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  public static ConnectingViewModel obtainViewModel(FragmentActivity activity) {
+    return new ViewModelProvider(activity).get(ConnectingViewModel.class);
+  }
+
+  private void openLaoDetailActivity(String laoId) {
+    Log.d(TAG, "Trying to open lao detail for lao with id " + laoId);
+
+    Intent intent = new Intent(this, LaoDetailActivity.class);
+    intent.putExtra("LAO_ID", laoId);
+    intent.putExtra("FRAGMENT_TO_OPEN", "LaoDetail");
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java:77]: Here also, it would be better to use constants.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/home/ConnectingFragmentTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+import dagger.hilt.android.testing.HiltAndroidRule;
+import dagger.hilt.android.testing.HiltAndroidTest;
+
+@HiltAndroidTest
+@RunWith(AndroidJUnit4.class)
+public class ConnectingFragmentTest {
+  private static final String laoName = "Lao Name";
+  private static final PublicKey pk = Base64DataUtils.generatePublicKey();
+  private static final Lao lao = new Lao(laoName, pk, 10223421);
+  private static final String laoId = lao.getId();
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/home/ConnectingFragmentTest.java:41]: In Java, we name constants with UPPERCASE and in snake_case.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/home/ConnectingFragmentTest.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  public void dynamicTextTest() {
+    laoConnectingText().check(matches(withText(laoId)));
+  }
+
+  @Test
+  public void cancelButtonLaunchesHomeTest() {
+    Intents.init();
+    cancelButton().perform(click());
+    intended(hasComponent(HomeActivity.class.getName()));
+  }
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/home/ConnectingFragmentTest.java:84]: I have not used much of Intents when testing, but I think you should use something along the line of `Intents.release()` at the end of the test.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/home/WalletFragmentTest.java [hunk 11]
```diff
@@ -41,7 +40,6 @@
 import dagger.hilt.android.testing.HiltAndroidRule;
 import dagger.hilt.android.testing.HiltAndroidTest;
 
-@LargeTest
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/ui/home/WalletFragmentTest.java:44]: Why did you remove this ?

**Review by GabrielFleischer — CHANGES_REQUESTED** [review]
Great work fixing this bug ! 👍 
I have some question here and there and there are some unintuitive things in the codebase that you might break.

So I made a lot of comments, don't hesitate to reply if something seems wrong.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java (hunk 6 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:322]: Sure, I lazily followed was already there. I've updated everything

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java (hunk 5 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java:1184]: I forgot to remove it sorry

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/MessageSender.java (hunk 4 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/MessageSender.java:64]: I add a Listener to avoid for the subscribe and catchup process to begin before the change of connection is finished

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java (hunk 2 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java:114]: My mistake, I misunderstood the purpose of onError

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java (hunk 7 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingActivity.java:34]: It does not like it at all if remove them :) 
Alternatively I could do something cleaner like

 `(String) Objects.requireNonNull(getIntent().getExtras()).get(getString(R.string.lao_id)))`

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailActivity.java [hunk 12]
```diff
@@ -60,8 +60,11 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
     setContentView(R.layout.lao_detail_activity);
     mViewModel = obtainViewModel(this);
     mViewModel.subscribeToLao(
-        (String) Objects.requireNonNull(getIntent().getExtras()).get("LAO_ID"));
-    if (getIntent().getExtras().get("FRAGMENT_TO_OPEN").equals("LaoDetail")) {
+        (String) Objects.requireNonNull(getIntent().getExtras()).get(getString(R.string.lao_id)));
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailActivity.java:63]: The `string.xml` should be used for text that will be displayed to the user, and you should keep using constants defined in the code for things like keys and so.

https://stackoverflow.com/questions/16160047/android-strings-xml-vs-static-constants

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingViewModel.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+
+    Log.d(TAG, "connecting to lao " + channelId);
+    disposables.add(
+        networkManager
+            .getMessageSender()
+            .subscribe(lao.getChannel())
+            .subscribe(
+                () -> {
+                  Log.d(TAG, "subscribing to LAO with id " + lao.getId());
+                  Log.d(TAG, "got success result for subscribe to lao");
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/connecting/ConnectingViewModel.java:68]: Are the two logs really needed here ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LaoHandler.java [hunk 14]
```diff
@@ -41,7 +42,11 @@ public static void handleCreateLao(HandlerContext context, CreateLao createLao)
     Channel channel = context.getChannel();
 
     Log.d(TAG, "handleCreateLao: channel " + channel + ", msg=" + createLao);
-    Lao lao = laoRepository.getLaoByChannel(channel);
+    Lao lao = new Lao(createLao.getId());
+
+    // Adding the newly created LAO to the repository
+    laoRepository.getLaoById().put(lao.getId(), new LAOState(lao));
+    laoRepository.setAllLaoSubject();
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LaoHandler.java:49]: Nice fix !

**Review by GabrielFleischer — DISMISSED** [review]
Almost done to me, I just have a small issue with the way you define constants.

Usually, I like to see string.xml as the localizable strings -> what the user will see
And the rest of the constants as usual static final strings.

But this is a big nitpick so I let you fix it if you want and approve for now.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/Constants.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
+package com.github.dedis.popstellar.utility;
+
+public class Constants {
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/Constants.java:3]: Pure nitpicking, but why creating a whole new class ?
You could have defined it in the fragment classes.

**Review by GabrielFleischer — APPROVED** [review]
LGTM

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/Constants.java (hunk 15 above)
Comment (Warhedgehog) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/Constants.java:3]: Because they are extras passed across activities, I thought it was cleaner that way. Plus I thought it might be useful to add some other constants in the future
